### DISCUSSION TITLE
MsaView: how to remove gap stripping

### DISCUSSION THREAD

**Discussion by mamrosh** [GMOD/jbrowse-components#4304] [General]
Hi, I've been using MsaView and was wondering where in msa_plugin.js the code to strip gaps in alignments is. I don't want this done by default when loading an alignment.

Thank you (for this and all your past help)!

**Comment by cmdcolin** [comment]
hello, so you are saying there are columns that are completely filled by gaps and it is automatically collapsing them? I will have to review to see how to avoid this. there is logic that tries to allow collapsing gaps

if you have sample data feel free to send it to me. if it is automatically removing those columns, i might need to make a new setting to avoid toggle this on/off. 

also, there is a lot of work actually happening on the msaview right now. if you're interested in any previews or interested in a demo, let me know!

**Comment by mamrosh** [comment]
Thanks for the fast response! I definitely see the rationale for collapsing gaps when possible for most users of msaview. However, I have some circumstances where I have a reference sequence as a track, and the actual alignment just has gaps in a region. For example, see the below example stockholm "file":

#=GC reference    WNTLIWQDTENSTFWQKFY
seq1_alignment    WNTLIWQDTEN-TFWQKFY
seq2_alignment.  WNTLIWQDTEN-TFWQKFY

In msaview, the "S" no longer appears in the reference sequence, since gap stripping occurred. There is custom numbering that goes along with the reference sequence, which won't work correctly if gaps get removed. I have tried to figure out where in the code the gap stripping occurs but haven't found it yet. In the meantime, I have added the reference sequence to the alignment itself, but I don't really want it there long-term (since I want it as a track on the top, which is closer to other custom things I've added to msaview). 

I would definitely be interested in a preview or demo of updates to msaview. My group likely will release a web tool to the public in a few months using msaview but I could envision additional things we could use msaview for, too.

Thank you!

**Reply by cmdcolin** [reply]
if you want you can schedule an office hours here https://calendar.app.google/1AYZkNCQNmwdY2R26 or i can send you some slides

it would be good to hear your thoughts as the new version does not display coordinates by default now, as we were unsure how coordinates should be reflected with the collapsed columns, but we may want to add them back in at least optionally perhaps (or on a per-row basis)

**Reply by cmdcolin** [reply]
and just as an add-on note, I can probably make a fix for the specific issue you are requesting just on the version that you already have, i'd just have to check into it

**Reply by mamrosh** [reply]
I scheduled office hours for two weeks from now and am excited to see the msaview changes. If you are able to figure out where the gap stripping occurs in the code I'm currently using, that would be much appreciated. However, I'm sure I'll eventually figure it out if you're not able to look into it. Thanks!

**Comment by cmdcolin** [comment]
@mamrosh here is an alternative build that preserves gaps, it makes the "skipBlanks" function do nothing

it is built from react-msaview  v1.3.2 (jbrowse-plugin-msaview v1.0.18)
diff

```
+// special build, don't skip blanks RE: https://github.com/GMOD/jbrowse-components/discussions/4304
 function skipBlanks(blanks: number[], arg: string) {
-  let s = ''
-  let b = 0
-  for (let j = 0; j < arg.length; j++) {
-    if (j === blanks[b]) {
-      b++
-    } else {
-      s += arg[j]
-    }
-  }
-  return s
+  return arg
+  // let s = ''
+  // let b = 0
+  // for (let j = 0; j < arg.length; j++) {
+  //   if (j === blanks[b]) {
+  //     b++
+  //   } else {
+  //     s += arg[j]
+  //   }
+  // }
+  // return s
 }


```

an upload with the built UMD versions is attached

[build.tar.gz](https://github.com/GMOD/jbrowse-components/files/14707050/build.tar.gz)


when testing with the small stock file you sent, it preserves the gap (screenshot below)

![image](https://github.com/GMOD/jbrowse-components/assets/6511937/02383f4e-f494-4fe1-8761-ff59573bc7ab)

**Reply by cmdcolin** [reply]
not sure if i correctly have it setup to display the 'reference sequence' track above everything like in your demo but can look into that if there are any issues
